File: google-drive/typescript/src/services/comments.ts
/**
 * Comments service for Google Drive API.
 *
 * Manages comments on files.
 */

import {
  Comment,
  CommentList,
  CreateCommentRequest,
  UpdateCommentRequest,
  ListCommentsParams,
} from '../types';

/**
 * HTTP transport interface.
 */
export interface CommentTransport {
  request<T>(url: string, options: RequestOptions): Promise<T>;
}

/**
 * Request options.
 */
export interface RequestOptions {
  method?: string;
  body?: any;
  params?: Record<string, any>;
}

/**
 * Comments service interface.
 */
export interface CommentsService {
  /**
   * Create a new comment.
   *
   * @param fileId - File ID
   * @param request - Comment creation request
   * @returns The created comment
   */
  create(fileId: string, request: CreateCommentRequest): Promise<Comment>;

  /**
   * List comments for a file.
   *
   * @param fileId - File ID
   * @param params - Optional list parameters
   * @returns Comment list with pagination
   */
  list(fileId: string, params?: ListCommentsParams): Promise<CommentList>;

  /**
   * Get a specific comment.
   *
   * @param fileId - File ID
   * @param commentId - Comment ID
   * @returns The comment
   */
  get(fileId: string, commentId: string): Promise<Comment>;

  /**
   * Update a comment.
   *
   * @param fileId - File ID
   * @param commentId - Comment ID
   * @param request - Update request
   * @returns The updated comment
   */
  update(fileId: string, commentId: string, request: UpdateCommentRequest): Promise<Comment>;

  /**
   * Delete a comment.
   *
   * @param fileId - File ID
   * @param commentId - Comment ID
   */
  delete(fileId: string, commentId: string): Promise<void>;
}

/**
 * Implementation of CommentsService.
 */
export class CommentsServiceImpl implements CommentsService {
  private readonly baseUrl = 'https://www.googleapis.com/drive/v3';

  constructor(private transport: CommentTransport) {}

  async create(fileId: string, request: CreateCommentRequest): Promise<Comment> {
    return this.transport.request<Comment>(`${this.baseUrl}/files/${fileId}/comments`, {
      method: 'POST',
      body: request,
    });
  }

  async list(fileId: string, params?: ListCommentsParams): Promise<CommentList> {
    return this.transport.request<CommentList>(`${this.baseUrl}/files/${fileId}/comments`, {
      method: 'GET',
      params: params as any,
    });
  }

  async get(fileId: string, commentId: string): Promise<Comment> {
    return this.transport.request<Comment>(
      `${this.baseUrl}/files/${fileId}/comments/${commentId}`,
      {
        method: 'GET',
      }
    );
  }

  async update(fileId: string, commentId: string, request: UpdateCommentRequest): Promise<Comment> {
    return this.transport.request<Comment>(
      `${this.baseUrl}/files/${fileId}/comments/${commentId}`,
      {
        method: 'PATCH',
        body: request,
      }
    );
  }

  async delete(fileId: string, commentId: string): Promise<void> {
    await this.transport.request<void>(
      `${this.baseUrl}/files/${fileId}/comments/${commentId}`,
      {
        method: 'DELETE',
      }
    );
  }
}

/**
 * Mock implementation for testing.
 */
export class MockCommentsService implements CommentsService {
  private comments = new Map<string, Map<string, Comment>>();
  private nextId = 1;

  async create(fileId: string, request: CreateCommentRequest): Promise<Comment> {
    if (!this.comments.has(fileId)) {
      this.comments.set(fileId, new Map());
    }

    const now = new Date().toISOString();
    const comment: Comment = {
      kind: 'drive#comment',
      id: `comment-${this.nextId++}`,
      createdTime: now,
      modifiedTime: now,
      author: {
        kind: 'drive#user',
        displayName: 'Mock User',
        photoLink: '',
        me: true,
        permissionId: 'permission-1',
        emailAddress: 'user@example.com',
      },
      htmlContent: request.content,
      content: request.content,
      deleted: false,
      resolved: false,
      anchor: request.anchor,
      quotedFileContent: request.quotedFileContent,
      replies: [],
    };

    this.comments.get(fileId)!.set(comment.id, comment);
    return comment;
  }

  async list(fileId: string, params?: ListCommentsParams): Promise<CommentList> {
    const fileComments = this.comments.get(fileId);
    const comments = fileComments ? Array.from(fileComments.values()) : [];

    return {
      kind: 'drive#commentList',
      comments,
    };
  }

  async get(fileId: string, commentId: string): Promise<Comment> {
    const fileComments = this.comments.get(fileId);
    if (!fileComments) {
      throw new Error(`File not found: ${fileId}`);
    }

    const comment = fileComments.get(commentId);
    if (!comment) {
      throw new Error(`Comment not found: ${commentId}`);
    }

    return comment;
  }

  async update(fileId: string, commentId: string, request: UpdateCommentRequest): Promise<Comment> {
    const comment = await this.get(fileId, commentId);
    comment.content = request.content;
    comment.htmlContent = request.content;
    comment.modifiedTime = new Date().toISOString();
    return comment;
  }

  async delete(fileId: string, commentId: string): Promise<void> {
    const fileComments = this.comments.get(fileId);
    if (fileComments) {
      fileComments.delete(commentId);
    }
  }
}

/**
 * Create a mock comments service.
 */
export function createMockCommentsService(): CommentsService {
  return new MockCommentsService();
}

File: google-drive/rust/src/services/comments.rs
//! Comment operations service.

use crate::client::GoogleDriveClient;
use crate::error::GoogleDriveResult;
use crate::types::*;
use serde::{Deserialize, Serialize};

/// Service for comment operations.
pub struct CommentsService<'a> {
    client: &'a GoogleDriveClient,
}

impl<'a> CommentsService<'a> {
    /// Creates a new comments service.
    pub fn new(client: &'a GoogleDriveClient) -> Self {
        Self { client }
    }

    /// Creates a new comment.
    pub async fn create(
        &self,
        file_id: &str,
        request: CreateCommentRequest,
    ) -> GoogleDriveResult<Comment> {
        let path = format!("/files/{}/comments", file_id);
        self.client.post(&path, &request).await
    }

    /// Lists comments for a file.
    pub async fn list(
        &self,
        file_id: &str,
        params: Option<ListCommentsParams>,
    ) -> GoogleDriveResult<CommentList> {
        let path = format!("/files/{}/comments", file_id);
        if let Some(p) = params {
            self.client.get_with_params(&path, &p).await
        } else {
            self.client.get(&path).await
        }
    }

    /// Gets a specific comment.
    pub async fn get(
        &self,
        file_id: &str,
        comment_id: &str,
        params: Option<GetCommentParams>,
    ) -> GoogleDriveResult<Comment> {
        let path = format!("/files/{}/comments/{}", file_id, comment_id);
        if let Some(p) = params {
            self.client.get_with_params(&path, &p).await
        } else {
            self.client.get(&path).await
        }
    }

    /// Updates a comment.
    pub async fn update(
        &self,
        file_id: &str,
        comment_id: &str,
        request: UpdateCommentRequest,
    ) -> GoogleDriveResult<Comment> {
        let path = format!("/files/{}/comments/{}", file_id, comment_id);
        self.client.patch(&path, &request).await
    }

    /// Deletes a comment.
    pub async fn delete(&self, file_id: &str, comment_id: &str) -> GoogleDriveResult<()> {
        let path = format!("/files/{}/comments/{}", file_id, comment_id);
        self.client.delete(&path).await
    }
}

// Request/Response types

/// Request to create a comment.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CreateCommentRequest {
    pub content: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub anchor: Option<String>,
}

/// Parameters for listing comments.
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct ListCommentsParams {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub include_deleted: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub page_size: Option<u32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub page_token: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub start_modified_time: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub fields: Option<String>,
}

/// Parameters for getting a comment.
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct GetCommentParams {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub include_deleted: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub fields: Option<String>,
}

/// Request to update a comment.
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct UpdateCommentRequest {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub content: Option<String>,
}
